File: Laya/GPUSkinning/src/GPUSkinning/Datas/GPUSkinningBone.ts
import Matrix4x4 = Laya.Matrix4x4;
import ByteReadUtil from "./ByteReadUtil";
/** 烘焙动画--骨骼节点信息 */
export default class GPUSkinningBone
{
    /** 编号 */
    guid: string;

    /** 名称 */
    name: string;

    /** 是否导出 */
    isExposed: boolean;

    /** 矩阵 */
    bindpose: Matrix4x4;

    /** 索引 */
    boneIndex: int = -1;

    /** 父节点索引 */
    parentBoneIndex: int = -1;

    /** 子节点索引 */
    childrenBonesIndices: int[];

    /** 是否生成了逆矩阵 */
    private _bindposeInvInit = false;

    private _bindposeInv: Matrix4x4;
    /** 逆矩阵 */
    public get BindposeInv(): Matrix4x4
    {
        if(!this._bindposeInv)
        {
            this._bindposeInv = new Matrix4x4();
            this.bindpose.invert(this._bindposeInv);
            this._bindposeInvInit = true;
        }
        return this._bindposeInv;
    }
    
    FromBytes(data: ArrayBuffer): void
    {
        var b:Laya.Byte = new Laya.Byte(data);
        this.name = b.readUTFString();
        this.guid = b.readUTFString();
        this.isExposed = b.readByte() != 0;
        this.boneIndex = b.readInt32();
        this.parentBoneIndex = b.readInt32();
        this.bindpose = ByteReadUtil.ReadMatrix4x4(b);

    }

    static CreateFromBytes(data: ArrayBuffer):GPUSkinningBone
    {
        var obj = new GPUSkinningBone();
        obj.FromBytes(data);
        return obj;
    }


}

File: Unity3D/GPUSkinning/Assets/GPUSkinning/Scripts/GPUSkinningBone.cs
using UnityEngine;
using System.Collections;
using System.IO;

[System.Serializable]
public class GPUSkinningBone
{
	[System.NonSerialized]
	public Transform transform = null;

	public Matrix4x4 bindpose;

	public int parentBoneIndex = -1;

	public int[] childrenBonesIndices = null;

	[System.NonSerialized]
	public Matrix4x4 animationMatrix;

	public string name = null;

    public string guid = null; 

    public bool isExposed = false;

    [System.NonSerialized]
    private bool bindposeInvInit = false;
    [System.NonSerialized]
    private Matrix4x4 bindposeInv;
    public Matrix4x4 BindposeInv
    {
        get
        {
            if(!bindposeInvInit)
            {
                bindposeInv = bindpose.inverse;
                bindposeInvInit = true;
            }
            return bindposeInv;
        }
    }


    public static GPUSkinningBone CreateFromBytes(byte[] bytes)
    {
        GPUSkinningBone obj = new GPUSkinningBone();
        obj.FromBytes(bytes);
        return obj;
    }

    public void FromBytes(byte[] bytes)
    {
    }


    public MemoryStream ToSteam()
    {
        MemoryStream stream = new MemoryStream();
        BinaryWriter b = new BinaryWriter(stream);
        b.WriteUTFString(name);
        b.WriteUTFString(guid);
        b.Write((bool) isExposed);
        b.Write((int) parentBoneIndex);
        b.WriteMatrix4x4(bindpose);

        return stream;
    }
}
